wp-includes/class-wp-offline-page-ui.php

```diff
@@ -0,0 +1,327 @@
+	 * Renders the settings section.
+	 */
+	public function render_settings() {
+		$has_pages = $this->has_pages();
+
+		if ( $has_pages ) :
+			?>
+			<label for="<?php echo esc_attr( WP_Offline_Page::OPTION_NAME ); ?>">
+				<?php esc_html_e( 'Select an existing page:', 'pwa' ); ?>
+			</label>
+			<?php
+			$this->render_page_dropdown();
+
+			$create_text = __( 'create a new one', 'pwa' );
+		else :
+			$create_text = __( 'Create a new offline page', 'pwa' );
+		endif;
+		?>
+		<p>
+			<span>
+				<?php
+				if ( $has_pages ) {
+					esc_html_e( 'or ', 'pwa' );
+				}
+				?>
+				<a href="<?php echo esc_url( admin_url( 'options-reading.php?action=create-offline-page' ) ); ?>"><?php echo esc_html( $create_text ); ?></a>.
+			</span>
+		</p>
+		<p class="description"><?php esc_html_e( 'This page is for the Progressive Web App (PWA), providing a default offline page. It is similar to a 404 page.  Unlike a 404 page that displays when the content is not found, this offline page shows when the person\'s internet is down, such as during a flight, or intermittent, such as going through tunnel.', 'pwa' ); ?></p>
+		<?php
+	}
+
+	/**
+	 * Renders the page dropdown.
+	 */
+	protected function render_page_dropdown() {
+		$non_offline_static_pages = $this->manager->get_static_pages();
+		unset( $non_offline_static_pages[ WP_Offline_Page::OPTION_NAME ] );
+
+		wp_dropdown_pages(
+			array(
+				'name'              => esc_html( WP_Offline_Page::OPTION_NAME ),
+				/* Translators: %1$s: A long dash */
+				'show_option_none'  => sprintf( esc_html__( '%1$s Select %1$s', 'pwa' ), '&mdash;' ),
+				'option_none_value' => '0',
+				'selected'          => intval( $this->manager->get_offline_page_id() ),
+				'post_status'       => array( 'draft', 'publish' ),
+				'exclude'           => implode( ',', $non_offline_static_pages ), // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped,WordPress.XSS.EscapeOutput.OutputNotEscaped -- This is a false positive.
+			)
+		);
+	}
+
+	/**
+	 * Handle the create-offline-page admin action.
+	 *
+	 * Will redirect to the newly created post upon success (and thus will not return).
+	 */
+	public function handle_create_offline_page_action() {
+
+		// Bail out if this is not the right page.
+		$screen = get_current_screen();
+		if ( ! $screen instanceof WP_Screen || 'options-reading' !== $screen->id ) {
+			return;
+		}
+
+		$r = $this->create_new_page();
+		if ( is_wp_error( $r ) ) {
+			add_settings_error(
+				WP_Offline_Page::OPTION_NAME,
+				WP_Offline_Page::OPTION_NAME,
+				__( 'Unable to create the offline page.', 'pwa' ),
+				'error'
+			);
+		} else {
+			wp_safe_redirect( get_edit_post_link( $r, 'raw' ) );
+			exit;
+		}
+	}
+
+	/**
+	 * Creates a new offline page.
+	 *
+	 * @return int|WP_Error The page ID or WP_Error on failure.
+	 */
+	public function create_new_page() {
+
+		$page_id = wp_insert_post(
+			wp_slash( array(
+				'post_title'   => __( 'Offline', 'pwa' ),
+				'post_status'  => 'draft',
```

Comment:
As discussed, let's have this create a `publish`'ed page. If we allow selecting a draft page then it wouldn't be accessible to the user.